src/core/lib/event_engine/workqueue.h

```diff
@@ -0,0 +1,238 @@
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryLockAndPop(bool front) ABSL_LOCKS_EXCLUDED(mu_) {
+    // Do not block the worker if there are other workers trying to pop
+    // tasks from this queue.
+    if (!mu_.TryLock()) return absl::nullopt;
+    auto mu_cleanup = absl::MakeCleanup([this]() {
+      mu_.AssertHeld();
+      mu_.Unlock();
+    });
+    if (GPR_UNLIKELY(elements_.empty())) {
+      if (most_recent_element_enqueue_timestamp_.load(
+              std::memory_order_relaxed) == kInvalidTimestamp) {
+        return absl::nullopt;
+      }
+      if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+      absl::optional<T> ret = absl::nullopt;
+      if (GPR_LIKELY(most_recent_element_.has_value())) {
+        most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                     std::memory_order_relaxed);
+        ret = std::exchange(most_recent_element_, absl::nullopt);
+      }
+      most_recent_element_lock_.Unlock();
+      return ret;
+    }
+    // the queue has elements, let's pop one and update timestamps
+    Storage ret_s;
+    if (front) {
+      ret_s = std::move(elements_.front());
+      elements_.pop_front();
+    } else {
+      ret_s = std::move(elements_.back());
+      elements_.pop_back();
+    }
+    if (elements_.empty()) {
+      oldest_enqueued_timestamp_.store(kInvalidTimestamp,
+                                       std::memory_order_relaxed);
+    } else if (front) {
+      oldest_enqueued_timestamp_.store(elements_.front().enqueued(),
+                                       std::memory_order_relaxed);
+    }
+    return ret_s.TakeElement();
+  }
+
+  // Attempts to pop from the back of the queue (most recent).
+  // This will return nullopt if the queue is empty, or if other workers
+  // are already attempting to pop from this queue.
+  absl::optional<T> TryPopMostRecentElement() {
+    if (!most_recent_element_lock_.TryLock()) return absl::nullopt;
+    if (GPR_UNLIKELY(!most_recent_element_.has_value())) {
+      most_recent_element_lock_.Unlock();
+      return absl::nullopt;
+    }
+    most_recent_element_enqueue_timestamp_.store(kInvalidTimestamp,
+                                                 std::memory_order_relaxed);
+    absl::optional<T> tmp = std::exchange(most_recent_element_, absl::nullopt);
+    most_recent_element_lock_.Unlock();
+    return tmp;
+  }
+
+  // The managed items in the queue
+  std::deque<Storage> elements_ ABSL_GUARDED_BY(mu_);
+  // The most recently enqueued element. This is reserved from work stealing
+  absl::optional<T> most_recent_element_
+      ABSL_GUARDED_BY(most_recent_element_lock_);
+  absl::base_internal::SpinLock ABSL_ACQUIRED_AFTER(mu_)
+      most_recent_element_lock_;
+  // TODO(hork): consider ABSL_CACHELINE_ALIGNED
+  std::atomic<grpc_core::Timestamp> most_recent_element_enqueue_timestamp_{
```

Comment:
this should definitely be `std::atomic<int64_t>` - there's no specialization for Timestamp, so this will end up adding a mutex

@soheilhy I'm assuming this is meant for some kind of fairness ordering, and on that assumption likely the millisecond granularity provided by `Timestamp` isn't right

Reply:
It should be an `std::atomic<int64_t>`, thanks. I'll use millis for it.

This variable is mostly being used as an inexpensive check to see if the most-recent-element is populated, an atomic bool could do that job. The one time that its Timestamp value is actually used is when populating the oldest enqueued timestamp. And that feature is useful for gathering stats, and potentially for smarter auto-scaling heuristics.

Reply:
Using atomic<int64_t> made no significant performance difference in my tests (see the last 2 commits with benchmark results).

Reply:
Let's keep the int64_t atomic as you suggested, it may be contributing to the opt-build performance wins we're seeing now.